### ISSUE TITLE
Parsing fundamentaly broken

### ISSUE THREAD

**Issue by lucaswerkmeister** [MatrixTeo/GoodCalc#15]
> `2×3^2`
> 36 // should be 18
> `1+2^3+4`
> 31 // should be 13
> `(2)^5`
> 7 // should be 32
> `2^(5)`
> 7 // should be 32
> `(1)!`
> // error
> `π!`
> // error
> `010`
> 8 // should be 10, but in JS number literals beginning with 0 are octal

**Parsing arithmetic expressions with regular expressions is fundamentally wrong.** Arithmetical expressions with correct operator precedence are a context-free language, which classical regular expressions *cannot* parse. It may be possible to do with extended regexes (see [this SO answer](http://stackoverflow.com/a/1732454/1420237)), but it will inevitably become an utterly unmaintainable mess of unreadable regexes. I strongly urge you to write an actual parser, and evaluate the expression yourself rather than abusing `eval`.

**Comment by MatrixTeo** [comment]
Totally agree with you, a parser that builds a tree is actually more robust and mantainable...
I've just discovered that eval() is also against the [content security policy](https://developer.mozilla.org/en-US/Apps/Build/Building_apps_for_Firefox_OS/CSP).

I'll consider this option for the future, although requires a lot of work :)

**Comment by lucaswerkmeister** [comment]
> I've just discovered that eval() is also against the content security policy.

Does that mean that the app can no longer be published to the Marketplace?

**Comment by MatrixTeo** [comment]
I don't think so, the CSP is applied only to "privileged apps" (apps with more permissions).
